File: NeetCode/Arrays_and_Hashing/Count_Vowel_String_In_Ranges/count__vowel__string__in__ranges.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def vowelStrings(self, words: List[str], queries: List[List[int]]) -> List[int]:
        vowels = set("aeiou")

        dp = [0]
        for word in words:
            word = word.lower()
            if word[0] in vowels and word[-1] in vowels:
                dp.append(dp[-1] + 1)
            else:
                dp.append(dp[-1])

        result = []

        for left, right in queries:
            result.append(dp[right + 1] - dp[left])
        return result
        pass


if __name__ == "__main__":
    sol = Solution()
    words = ["aba", "bcb", "ece", "aa", "e"]
    queries = [[0, 2], [1, 4], [1, 1]]
    print(sol.vowelStrings(words, queries))
    print("Running Solution...")

File: NeetCode/Arrays_and_Hashing/Count_Vowel_String_In_Ranges/Count_Vowel_String_In_Ranges.cpp
// Count_Vowel_String_In_Ranges.cpp
//
// Compile locally with:
// clang++ Count_Vowel_String_In_Ranges.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<int> vowelStrings(vector<string> &words, vector<vector<int>> &queries)
    {
        unordered_set<char> vowels = {'a', 'e', 'i', 'o', 'u'};
        vector<int> dp = {0};

        for (auto &word : words)
        {
            int valid = 0;
            if (vowels.count(word[0]) > 0 && vowels.count(word[word.size() - 1]) > 0)
            {
                valid = 1;
            }
            dp.push_back(dp.back() + valid);
        }

        vector<int> results;

        for (auto &tmp : queries)
        {
            int left = tmp[0];
            int right = tmp[1];

            results.push_back(dp[right + 1] - dp[left]);
        }
        return results;
    }
};

int main()
{
    cout << "Running Count_Vowel_String_In_Ranges.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
